src/language/isolates.md

```diff
@@ -0,0 +1,1106 @@
+- Finally, return the completer’s future, which will eventually contain the
+  response from the worker isolate.
+
+<?code-excerpt "lib/robust_ports_example/step_5_add_completers.dart (parse-json)"?>
+```dart
+Future<Object?> parseJson(String message) async {
+  final completer = Completer<Object?>.sync();
+  final id = _idCounter++;
+  _activeRequests[id] = completer;
+  _commands.send((id, message));
+  return await completer.future;
+}
+```
+
+You also need to update `_handleResponsesFromIsolate`
+and `_handleCommandsToIsolate` to handle this system.
+
+In `_handleCommandsToIsolate`, you need to account for the `message` being a
+record with two values, rather than just the json text. Do so by destructuring
+the values from `message`.
+
+Then, after decoding the json, update the call to `sendPort.send` to pass both
+the id and the decoded json back to the main isolate, again using a record.
+
+<?code-excerpt "lib/robust_ports_example/step_5_add_completers.dart (handle-commands)"?>
+```dart
+static void _handleCommandsToIsolate(
+    ReceivePort receivePort, SendPort sendPort) {
+  receivePort.listen((message) {
+    final (int id, String jsonText) = message as (int, String); // New
+    try {
+      final jsonData = jsonDecode(jsonText);
+      sendPort.send((id, jsonData)); // Updated
+    } catch (e) {
+      sendPort.send((id, RemoteError(e.toString(), '')));
+    }
+  });
+}
+```
+
+Finally, update the `_handleResponsesFromIsolate`.
+
+- First, destructure the id and the response from the message argument again.
+- Then, remove the completer that corresponds to this request from
+  the `_activeRequests` map.
+- Lastly, rather than throwing an error or printing the decoded json, complete
+  the completer, passing in the response. When this completes, the response will
+  be returned to the code that called `parseJson` on the main isolate.
+
+<?code-excerpt "lib/robust_ports_example/step_5_add_completers.dart (handle-response)"?>
+```dart
+void _handleResponsesFromIsolate(dynamic message) {
+  final (int id, Object? response) = message as (int, Object?); // New
+  final completer = _activeRequests.remove(id)!; // New
+
+  if (response is RemoteError) {
+    completer.completeError(response); // Updated
+  } else {
+    completer.complete(response); // Updated
+  }
+}
+```
+
+#### Step 6: Add functionality to close the ports
+
+When the isolate is no longer being used by your code, you should close the
+ports on the main isolate and the worker isolate.
+
+First, add a class-level boolean that tracks if the ports are closed.
+
+Then, add the `Worker.close` method. Within this method:
+Update `_closed` to be true. Send a final message to the worker isolate. This
+message is a `String` that reads “shutdown”. This isn’t a specific string, and
+it could be any object you’d like. It will be used in the next code snippet.
+Finally, check if `_activeRequests` is empty. If it is, close down the main
+isolate’s `ReceivePort` called `_responses`
```

Comment:
```suggestion
- First, add a class-level boolean that tracks if the ports are closed.
- Then, add the `Worker.close` method. Within this method:
  - Update `_closed` to be true.
  - Send a final message to the worker isolate.
    This message is a `String` that reads “shutdown”,
    but it could be any object you’d like.
    You will use it in the next code snippet.
- Finally, check if `_activeRequests` is empty. If it is, close down the main
  isolate’s `ReceivePort` named `_responses`.
```
The long paragraph is formatted kind of weird, so I made the whole section match the rest of the sections (double check that the sub-bullet layers render properly)